pkg/serviceaccount/jwt.go

```diff
@@ -53,43 +56,143 @@ type TokenGenerator interface {
 
 // JWTTokenGenerator returns a TokenGenerator that generates signed JWT tokens, using the given privateKey.
 // privateKey is a PEM-encoded byte array of a private RSA key.
-// JWTTokenAuthenticator()
 func JWTTokenGenerator(iss string, privateKey interface{}) (TokenGenerator, error) {
-	var alg jose.SignatureAlgorithm
+	var signer jose.Signer
+	var err error
 	switch pk := privateKey.(type) {
 	case *rsa.PrivateKey:
-		alg = jose.RS256
+		signer, err = signerFromRSAPrivateKey(pk)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not generate signer for RSA keypair")
+		}
 	case *ecdsa.PrivateKey:
-		switch pk.Curve {
-		case elliptic.P256():
-			alg = jose.ES256
-		case elliptic.P384():
-			alg = jose.ES384
-		case elliptic.P521():
-			alg = jose.ES512
-		default:
-			return nil, fmt.Errorf("unknown private key curve, must be 256, 384, or 521")
+		signer, err = signerFromECDSAPrivateKey(pk)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not generate signer for ECDSA keypair")
 		}
 	case jose.OpaqueSigner:
-		alg = jose.SignatureAlgorithm(pk.Public().Algorithm)
+		signer, err = signerFromOpaqueSigner(pk)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not generate signer for OpaqueSigner")
+		}
 	default:
 		return nil, fmt.Errorf("unknown private key type %T, must be *rsa.PrivateKey, *ecdsa.PrivateKey, or jose.OpaqueSigner", privateKey)
 	}
 
+	return &jwtTokenGenerator{
+		iss:    iss,
+		signer: signer,
+	}, nil
+}
+
+func signerFromRSAPrivateKey(keyPair *rsa.PrivateKey) (jose.Signer, error) {
+	// Derive the KeyID from the public key, in a non-reversible way.  The
+	// KeyID helps OIDC verifiers cope with key rotations.  Making the
+	// derivation non-reversible makes it impossible for someone to
+	// accidentally obtain the real key from the key ID and use it for token
+	// validation.
```

Comment:
Even though the `KeyID` is generated using a one-way hash of the public key, is it useful for bad actors to collect JWTs they can assert were signed with the same private key?

Reply:
It seems like that's already possible, with some effort &mdash; the cluster advertises its public keys, and it's easy to test that a given JWT was signed by a given key.

Reply:
Oh, interesting. Thanks for the additional information.